src/rust.rs

```diff
@@ -707,34 +768,104 @@ impl<T> MutableHandle<T> {
     }
 }
 
-impl<T> Deref for MutableHandle<T> {
+impl<'a, T> MutableHandle<'a, T> {
+    pub unsafe fn from_marked_location(ptr: *mut T) -> Self {
+        RawMutableHandle::from_marked_location(ptr).into()
+    }
+
+    pub fn handle(&self) -> RawHandle<T> {
+        unsafe {
+            RawHandle::from_marked_location(self.ptr as *const T)
+        }
+    }
+
+    pub fn new(ptr: &'a mut T) -> Self {
+        Self { ptr: ptr }
+    }
+
+    pub fn get(&self) -> T
+        where T: Copy
+    {
+        *self.ptr
+    }
+
+    pub fn set(&mut self, v: T)
+        where T: Copy
+    {
+        *self.ptr = v
+    }
+
+    pub fn raw(&mut self) -> RawMutableHandle<T> {
+        unsafe {
+            RawMutableHandle::from_marked_location(self.ptr)
+        }
+    }
+}
+
+impl<T> Deref for RawMutableHandle<T> {
     type Target = T;
 
     fn deref<'a>(&'a self) -> &'a T {
         unsafe { &*self.ptr }
     }
 }
 
-impl<T> DerefMut for MutableHandle<T> {
+impl<T> DerefMut for RawMutableHandle<T> {
     fn deref_mut<'a>(&'a mut self) -> &'a mut T {
         unsafe { &mut *self.ptr }
     }
 }
 
-impl HandleValue {
-    pub fn null() -> HandleValue {
+impl<'a, T> Deref for MutableHandle<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        self.ptr
+    }
+}
+
+impl<'a, T> DerefMut for MutableHandle<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.ptr
+    }
+}
+
+impl<'a, T> From<MutableHandle<'a, T>> for RawMutableHandle<T> {
+    fn from(handle: MutableHandle<'a, T>) -> Self {
+        unsafe { RawMutableHandle::from_marked_location(handle.ptr) }
+    }
+}
+
+impl<'a, T> From<RawMutableHandle<T>> for MutableHandle<'a, T> {
```

Comment:
How important are the `From<Raw>` impls for the safe handles? It makes me uncomfortable that we can create safe handles with 'static lifetimes from arbitrary handles that do not actually have static lifetimes.

Reply:
removed.